scipy/stats/unuran/unuran_wrapper.pyx.templ

```diff
@@ -0,0 +1,869 @@
+        if N == 0:
+            return np.asarray(out).reshape(oshape)
+        self._ppf_hat(&u_view[0], &out[0], N)
+        if u.ndim == 0:
+            return out[0]
+        return np.asarray(out).reshape(oshape)
+
+
+cdef class DiscreteAliasUrn(Method):
+    r"""
+    Discrete Alias-Urn Method.
+
+    This method is used to sample from discrete distributions with a finite
+    domain. It uses the probability vector of size :math:`N` or a probability
+    mass function with a finite support to generator random numbers from the
+    distribution.
+
+    Parameters
+    ----------
+    pv_or_pmf : array_like, callable
+        Probability vector or the PMF of the distribution. The signature of
+        the PMF is expected to be: ``def pmf(k: int, *params) -> float``. i.e
+        the PMF should accecpt a Python integer and return a Python float.
+    params : tuple, optional
+        Parameters that the PMF takes. If a probability vector is available,
+        this parameter is ignored. Default is an empty tuple.
+    domain : int, optional
+        Support of the PMF. If a probability vector is not available, a
+        finite domain must be given. i.e. the PMF must have a finite support.
+    urnfactor : float, optional
+        Size of the urn table *relative* to the size of the probability
+        vector. It must not be less than 1. Larger tables result is faster
+        generation times but require a more expensive setup.
+    seed : int, BitGenerator, Generator, RandomState, SeedSequence, optional
+        Seed for the underlying uniform random number generation.
+
+    Notes
+    -----
+    This method works when either a finite probability vector is available or
+    the PMF of the distribution is available. In case a PMF is only available,
+    the support (domain) of the PMF must also be given. It is recommended to
+    first obtain the probability vector be evaluating the PMF at each point
+    in the support and then using the probability vector instead (because this
+    method accepts arbitrarily sized probability vectors but not arbitrarily
+    sized domains).
+
+    If a probability vector is given, it must be a 1-dimensional array of
+    non-negative floats without any ``inf`` or ``nan`` values. Also, there
+    must be at least one non-zero entry otherwise an exception is raised.
+
+    The parameter ``unrfactor`` can be increased for faster generation at the
+    cost of increased setup time. This method uses a table for random
+    variate generation. ``unfactor`` controls the size of this table relative
+    to the size of the probability vector (or width of the support, in case a
+    PV is not available). As this table is computed during setup time,
+    increasing this parameter linearly increases the time required to setup.
+    It is recommended to keep this parameter under 2.
+
+    References
+    ----------
+    .. [1] UNU.RAN reference manual, Section 5.8.2,
+           "DAU - (Discrete) Alias-Urn method",
```

Comment:
Please add reference WAa77 given in the unu.ran doc

Reply:
Done.